errors/container.go

```diff
@@ -0,0 +1,185 @@
+// The Error Container entity serves a purpose for keeping track of errors,
+// details and fields information. This component can be used explicitly (for
+// example when we want to sequentially fill it with details and fields), as
+// well as implicitly (all errors that are returned from handler are
+// transformed to an Error Container, and passed as GRPCStatus to a gRPC Gateway).
+package errors
+
+import (
+	"fmt"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/golang/protobuf/proto"
+
+	"github.com/infobloxopen/atlas-app-toolkit/rpc/errdetails"
+	"github.com/infobloxopen/atlas-app-toolkit/rpc/errfields"
+)
+
+// Container struct is an entity that servers a purpose of error container and
+// consist of methods to append details, field errors and setting general
+// error code/message.
+type Container struct {
+	// details field contains an array of error details.
+	details []*errdetails.TargetInfo
+
+	// fields field contains per-field error map.
+	fields *errfields.FieldInfo
+
+	// errCode, errMessage field contain the general error message.
+	errCode    codes.Code
+	errMessage string
+
+	// errSet flag indicates whether the error was set by calling one of
+	// following methods: Set, WithDetail(s), WithField(s).
+	errSet bool
```

Comment:
Can you give more details why do we need this flag?

Reply:
we have no way to detect whether the error was set to 'Unknown' or not. it's used with methods Set/IfSet/WithDetails/WithFields when we want to gather multiple errors throughout the function call.

```
func validateSmth(ctx context.Context, input ...) error {

    if someValidationFailed() {
        errors.Set(ctx, /* validation error */).WithDetails(/* details ... */)
    }

    if someOtherValidationFailed() {
        errros.Set(ctx, ...)
    }

    return errors.IfSet(ctx, /* ... general error msg */)
}
```

As you can see, we need to flag to indicate whether error was set at all.

In an alternative scenario, when we want to return error immediately we would write something like that (errSet flag is not needed):

func validateSmth2(ctx context.Context, ...) error {
     if someValidateionFailed() {
          return New(ctx, /* err msg */).WithDetails(/* field details */).WithFields(/* fields info */)
     }

     return nil
}